### ISSUE TITLE
update-available detected but not downloading with Amazon S3

### ISSUE THREAD

**Issue by sudhamab** [electron-userland/electron-builder#4810]
* **Version**: ^21.2.0

* **Version**:  ^4.2.2

* **Target**: Mac

Hi Team, 

Thanks for this great library and hope you are all safe at home during these tough times. 

We have been trying very hard to make the auto update work . First we tried with github private repo and updates were not getting detected. 

Then we switched to Amazon S3. The update-available event gets triggered but the latest update doesn't seem to be downloading. We are not getting any error either. 

The "download-progress" event does not  triggered. 

The code for main.js is as follows 

```
const { app, BrowserWindow, ipcMain, powerMonitor } = require("electron");
const { autoUpdater } = require("electron-updater");
const log = require("electron-log");
const isDev = require("electron-is-dev");


let mainWindow, authWindow, zoomAuthWindow;

function createWindow() {
	const startUrl =
		process.env.ELECTRON_START_URL ||
		url.format({
			pathname: path.join(__dirname, "../index.html"),
			protocol: "file:",
			slashes: true
		});
		
	mainWindow = new BrowserWindow({
		width: 1400,
		height: 800,
		webPreferences: {
			preload: path.join(__dirname, "preload.js"),
			nodeIntegration: true
		}
	});

	console.log("URL", startUrl);
	mainWindow.loadURL(startUrl);
	mainWindow.on("closed", function() {
		mainWindow = null;
	});

	if (isDev) {
		log.info("running in dev");
	} else {
		log.info("running in PROD");
		autoUpdater.checkForUpdatesAndNotify();
	}

}

app.on("ready", () => {
	createWindow();
});

ipcMain.on("app_version", event => {
	log.info("received event ", app.getVersion());
	event.sender.send("app_version", { version: app.getVersion() });
});

autoUpdater.on("checking-for-update", () => {
	log.info("checking for updates");
});

autoUpdater.on("update-available", () => {
	log.info("auto updater sending update available");
	mainWindow.webContents.send("update_available");
});

autoUpdater.on("download-progress", progressObj => {
	log.info("Tracking progress");
	var log_message = "Download speed: " + progressObj.bytesPerSecond;
	log_message = log_message + " - Downloaded " + progressObj.percent + "%";
	log_message =
		log_message +
		" (" +
		progressObj.transferred +
		"/" +
		progressObj.total +
		")";
	log.info(log_message);
	mainWindow.webContents.send("download_progress", log_message);
});

autoUpdater.on("update-not-available", () => {
	log.info("auto updater sending update NOT available");
	mainWindow.webContents.send("update_not_available");
});

autoUpdater.on("update-downloaded", () => {
	log.info("auto updater sending update downloaded");
	mainWindow.webContents.send("update_downloaded");
});

ipcMain.on("restart_app", () => {
	autoUpdater.quitAndInstall();
});
```

The index.html file is as follows

```
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <link rel="icon" href="%PUBLIC_URL%/favicon.ico" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="theme-color" content="#000000" />
  <meta name="description" content="Web site created using create-react-app" />
  <link rel="apple-touch-icon" href="logo192.png" />

  <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
  
  <title>React App</title>
  <style>
    #notification {
      position: fixed;
      bottom: 20px;
      left: 20px;
      width: 200px;
      padding: 20px;
      border-radius: 5px;
      background-color: white;
      box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
      z-index: 3002
    }

    .hidden {
      display: none;
    }
  </style>
</head>


<body>
  <noscript>You need to enable JavaScript to run this app.</noscript>
  <div id="root">
  </div>
  <div id="notification" class="hidden">
    <p id="version"></p>
    <p id="message"></p>
    <button id="close-button" onClick="closeNotification()">
      Close
    </button>
    <button id="restart-button" onClick="restartApp()" class="hidden">
      Restart
    </button>
  </div>
  <script>
    const {
      ipcRenderer
    } = require('electron');
    const log = require("electron-log");

    const version = document.getElementById('version');
    const notification = document.getElementById('notification');
    const message = document.getElementById('message');
    const restartButton = document.getElementById('restart-button');

    ipcRenderer.send('app_version');
    ipcRenderer.on('app_version', (event, arg) => {
      log.info('ipcRenderer on app_version')
      ipcRenderer.removeAllListeners('app_version');
      version.innerText = 'Version ' + arg.version;
    });

    ipcRenderer.on('update_not_available', () => {
      log.info('ipcRenderer on update_not_available')
      ipcRenderer.removeAllListeners('update_not_available');
      message.innerText = 'NO update is available';
      notification.classList.remove('hidden');
    });

    ipcRenderer.on('update_available', () => {
      log.info('ipcRenderer on update_available')
      ipcRenderer.removeAllListeners('update_available');
      message.innerText = 'A new update is available. Downloading now...';
      notification.classList.remove('hidden');
    });

    ipcRenderer.on('download_progress', (progress_message) => {
      log.info('ipcRenderer on download_progress')
      ipcRenderer.removeAllListeners('update_available');
      message.innerText = progress_message;
      notification.classList.remove('hidden');
    });

    ipcRenderer.on('update_downloaded', () => {
      log.info('ipcRenderer on update_downloaded')
      ipcRenderer.removeAllListeners('update_downloaded');
      message.innerText = 'Update Downloaded. It will be installed on restart. Restart now?';
      restartButton.classList.remove('hidden');
      notification.classList.remove('hidden');
    });

    function closeNotification() {
      notification.classList.add('hidden');
    }

    function restartApp() {
      ipcRenderer.send('restart_app');
    }
  </script>
</body>
</html>
```

The two main tutorials we followed were as follows 
[Tutorial 1](https://medium.com/@johndyer24/creating-and-deploying-an-auto-updating-electron-app-for-mac-and-windows-using-electron-builder-6a3982c0cee6)

[Tutorial 2](https://blog.brainio.com/development/automatic-updates-in-electron-using-amazon-s3/#Silent_updates_flow)

Would be grateful if you can help us identify what is going wrong as we need to release a version asap.

PS I didn't comment [inside this issue](https://github.com/electron-userland/electron-builder/issues/2377) in case this has anything to do with S3.

**Comment by sanzog03** [comment]
refer this official doc `https://www.electron.build/auto-update` instead.
tldr on the doc part (which is actually not that long) what is missing on your side is: `autoUpdater.on("update-available", () => {
	log.info("auto updater sending update available");
       //needs autoUpdater.downloadUpdates(); 
        autoUpdater.downloadUpdates();
	mainWindow.webContents.send("update_available");
});`
i.e. the code on your side only receives an update-available event and then shows it on renderer through IPC.
what you need to do is trigger a download initiation as well, either automatically - as done above - or through user input.
This can be done using `autoDownload = true` flag as well. This is written well in the docs.

**Comment by sudhamab** [comment]
thanks @sanzog03 ! I looked at the docs on the link you shared and also looked at the examples in it.  And the downloadUpdates() the way you showed also did not work for me but i will try again.

in the link  you shared above, the documentation for **appUpdater.checkForUpdatesAndNotify()**
method says 

> Asks the server whether there is an update, **download** and notify if update available

So i interpreted that as it will Download automatically as well. 

And if i use just the **autoUpdater.checkForUpdates()**
then i will have to use the autoUpdater.downloadUpdates()
So better to just use the **checkForUpdatesAndNotify** 
method. 

With that said, i also looked at the [example](https://gist.github.com/iffy/0ff845e8e3f59dbe7eaf2bf24443f104) which come from the link you sent 
and this too was not using the  autoUpdater.downloadUpdates(); inside. 

With that said, let me try this again with the combination of checkForUpdates and then downloadUpdates(). I may have done  the downloadUpdates along with checkForUpdatesAndNotify. Not sure if that makes a difference? 

Also, sorry for this lame question but just to avoid having to do two uploads, each which can take 20 minutes due to notarization etc, the autoDownload = true is to be passed as param in the event handler as below? 

```
autoUpdater.on("update-available", (autoUpdates = true) => {
	log.info("auto updater sending update available");
	mainWindow.webContents.send("update_available");
}); 
```

thanks again for your prompt help!

**Comment by sudhamab** [comment]
I replaced autoUpdater.checkForUpdatesAndNotify(); with autoUpdater.checkForUpdates();

and then the following 

```
autoUpdater.on("update-available", () => {
	log.info("auto updater sending update `available");`
	autoUpdater.downloadUpdates();
	mainWindow.webContents.send("update_available");
});
```

BUT it's not working. No progress being logged.

Did I misunderstand something?

**Comment by sanzog03** [comment]
try setting env `ELECTRON_ENABLE_LOGGING=true` before starting the electron application.
This will give you the detailed logs. helps to see the problem.

Using s3 to publish artifact has some configurations things to be looked at.
in the build config, is the `acl =  “private” or “public-read”` which is public-read by default.

If private, you need to pre-Sign the request-headers during latest version check, using aws4 or equivalent package.
else,
login to the s3 and in the release folder see if there is old binary and latest binary and a corresponding `latest-<xxx>.yml`. 

using s3 as a platform to publish artifact is a strenuous work.
I suggest you to use github for publishing artifact and using the above suggested code.

**Comment by sudhamab** [comment]
thanks @sanzog03. I tried ELECTRON_ENABLE_LOGGING=true but the application is starting from the Applications folder. Where should i see these logs? 

I am using electron-log lib and that also is logging the update-available message. Not anything after that 

As for S3. one of the tutorials i mentioned in my original post mentions how to set up S3. I am able to updload to S3 with electron-builder. Yes there might be older builds in the folder but the yml file latest-mac.yml has the latest version. I checked that. 

And i think autoUpdater knows from it that a new version is available and the version number

With github as i mentioned, i was using a private repo so it wasn't even detecting the latest build even though the upload was happening. 

Quite perplexed.

**Comment by sudhamab** [comment]
Also, even after fixing the typo in autoUpdater.downloadUpdates();  [it should be autoUpdater.downloadUpdate(), without the 's' ], it did not work :(

**Comment by sudhamab** [comment]
hi, just wanted to follow up. if someone can help with this. Are release is being held up

**Comment by christarnowski** [comment]
The logs should be located in `~/Library/Logs/YOUR_APP_NAME` by default if you're using `electron-log`.

Can you confirm that the files (`*.yml` etc.) are on S3? Is the bucket set to allow public read?

**Comment by sudhamab** [comment]
Thanks Krzysztof! I finally got it to work with a couple of unknowns ( 1 issue stated at the end)  

So yes, the logs were there as you stated. I turned on the public read on each flle manually apart from changing the bucket policy also to public read fully including for ACL.  

Also, after adding 

```
autoUpdater.on("error", (event, err) => {
	log.error(err);
});
```

I realized that it was also expecting a zip version of the dmg on S3. 
which I had removed this from my package.json for local runs to save space

Once I added this to the package.json as below, it all worked saving the error which i have pasted at the end
```
"mac": {
      "category": "public.app-category.productivity",
      "target": [
        "dmg",
        "zip"
      ],
```

Got some good offline help from Marc Parillo ( @nbcnc ). Can't tag him here unfortunately

**The 2 unknowns are** 

**1. I got one error after the download completed** 
`Error: ENOENT: no such file or directory, rename '/Users/xyz/Library/Application Support/Caches/electron-cra-example-updater/pending/temp-XYZ ABC-1.0.0-mac.zip' -> '/Users/xyz/Library/Application Support/Caches/electron-cra-example-updater/pending/XYZ ABC-1.0.0-mac.zip'
`
**2. While it was downloading, the autoUpdated("on-progress") seemed to get called twice almost simultaneously**
Not sure if it's because both dmg and zip are being downloaded simultaneously but would be good to know

However, I was able to restart the app and it loaded the latest version

**Comment by martinamunoz** [comment]
This bug still present on Electron Updater. First I use GitHub, the download progress event was fired. When i changed to S3, the event is not triggered, same code.
`"dependencies": {
    "axios": "^0.19.2",
    "cron": "^1.8.2",
    "date-and-time": "^0.14.0",
    "electron-ipc": "^0.3.0",
    "electron-is-dev": "^1.2.0",
    "electron-log": "^4.3.0",
    "electron-settings": "^3.2.0",
    "electron-squirrel-startup": "^1.0.0",
    "electron-updater": "^4.3.5",
    "ipc": "0.0.1",
    "nconf": "^0.10.0",
    "nedb-promises": "^4.0.3",
    "node-gzip": "^1.1.2",
    "zlib": "^1.0.5"
  },
  "devDependencies": {
    "electron-builder": "^22.9.1",
    "electron": "^8.1.0"
  },`
